bitcoin/src/p2p/mod.rs

```diff
@@ -244,31 +244,47 @@ impl FromStr for Magic {
         }
     }
 }
-impl From<Network> for Magic {
-    fn from(network: Network) -> Magic {
-        match network {
-            Network::Bitcoin => Magic::BITCOIN,
-            Network::Testnet(TestnetVersion::V3) => Magic::TESTNET,
-            Network::Testnet(TestnetVersion::V4) => Magic::TESTNET4,
-            Network::Signet => Magic::SIGNET,
-            Network::Regtest => Magic::REGTEST,
+
+macro_rules! generate_network_magic_conversion {
+    ($($network:pat => Magic::$magic:ident,)*) => {
+        impl From<Network> for Magic {
+            fn from(network: Network) -> Magic {
+                match network {
+                    $(
+                        $network => Magic::$magic,
+                    )*
+                }
+            }
+        }
+    };
+    ($(Magic::$magic:ident => $network:expr,)*) => {
+        impl TryFrom<Magic> for Network {
+            type Error = UnknownMagicError;
+            fn try_from(magic: Magic) -> Result<Self, Self::Error> {
+                match magic {
+                    $(
+                        Magic::$magic => Ok($network),
+                    )*
+                    _ => Err(UnknownMagicError(magic)),
+                }
+            }
         }
     }
 }
 
-impl TryFrom<Magic> for Network {
-    type Error = UnknownMagicError;
-
-    fn try_from(magic: Magic) -> Result<Self, Self::Error> {
-        match magic {
-            Magic::BITCOIN => Ok(Network::Bitcoin),
-            Magic::TESTNET => Ok(Network::Testnet(TestnetVersion::V3)),
-            Magic::TESTNET4 => Ok(Network::Testnet(TestnetVersion::V4)),
-            Magic::SIGNET => Ok(Network::Signet),
-            Magic::REGTEST => Ok(Network::Regtest),
-            _ => Err(UnknownMagicError(magic)),
-        }
-    }
+generate_network_magic_conversion! {
+    Network::Bitcoin => Magic::BITCOIN,
+    Network::Testnet(TestnetVersion::V3) => Magic::TESTNET,
+    Network::Testnet(TestnetVersion::V4) => Magic::TESTNET4,
+    Network::Signet => Magic::SIGNET,
+    Network::Regtest => Magic::REGTEST,
+}
+generate_network_magic_conversion! {
+    Magic::BITCOIN => Network::Bitcoin,
+    Magic::TESTNET => Network::Testnet(TestnetVersion::V3),
+    Magic::TESTNET4 => Network::Testnet(TestnetVersion::V4),
+    Magic::SIGNET => Network::Signet,
+    Magic::REGTEST => Network::Regtest,
 }
```

Comment:
The point of the macro wast o **not** duplicate these items but have them together so that calling the macro synchronizes the conversions.

Reply:
The reason for duplicate here is that the patten of the original unique macro cannot satisfy us, because the structure of 'Network' has changed, and it cannot match as before, so I split it into two patten in order to keep the macro.

Do you have any good suggestions here,
- Retain macros (will result in some duplicate)?
- Delete macro ?

Reply:
I think it should still work since the sides are same but maybe you'll have to change things like `ident` to `tt`. If you can't figure it out and nobody wants to try then it's better to delete the macro entirely.

Reply:
According to your idea, I tried a variety of optimization methods, `tt` was not too smooth, and finally I merged the two branches to avoid 'duplicate'.

Reply:
`Network::$network:ident` could also be changed to something like `Network::$network:ident $((TestnetVersion::$testnet_version:ident))?`. The advantage of using idents is it is allowed in patterns and expressions.

Reply:
I copy pasted the original macro and added the TestnetVersion enum from this PR:

<details>
<summary>diff</summary>

```diff
diff --git a/bitcoin/src/p2p/mod.rs b/bitcoin/src/p2p/mod.rs
index 1121b2fb..30bd8cdc 100644
--- a/bitcoin/src/p2p/mod.rs
+++ b/bitcoin/src/p2p/mod.rs
@@ -246,12 +246,12 @@ impl FromStr for Magic {
 }
 
 macro_rules! generate_network_magic_conversion {
-    ($($network_pat:pat => $magic:ident => $network_expr:expr,)*) => {
+    ($(Network::$network:ident$((TestnetVersion::$testnet_version:ident))? => Magic::$magic:ident,)*) => {
         impl From<Network> for Magic {
             fn from(network: Network) -> Magic {
                 match network {
                     $(
-                        $network_pat => Magic::$magic,
+                        Network::$network$((TestnetVersion::$testnet_version))? => Magic::$magic,
                     )*
                 }
             }
@@ -263,7 +263,7 @@ macro_rules! generate_network_magic_conversion {
             fn try_from(magic: Magic) -> Result<Self, Self::Error> {
                 match magic {
                     $(
-                        Magic::$magic => Ok($network_expr),
+                        Magic::$magic => Ok(Network::$network$((TestnetVersion::$testnet_version))?),
                     )*
                     _ => Err(UnknownMagicError(magic)),
                 }
@@ -274,11 +274,11 @@ macro_rules! generate_network_magic_conversion {
 // Generate conversion functions for all known networks.
 // `Network -> Magic` and `Magic -> Network`
 generate_network_magic_conversion! {
-    Network::Bitcoin => BITCOIN => Network::Bitcoin,
-    Network::Testnet(TestnetVersion::V3) => TESTNET => Network::Testnet(TestnetVersion::V3),
-    Network::Testnet(TestnetVersion::V4) => TESTNET4 => Network::Testnet(TestnetVersion::V4),
-    Network::Signet => SIGNET => Network::Signet,
-    Network::Regtest => REGTEST => Network::Regtest,
+    Network::Bitcoin => Magic::BITCOIN,
+    Network::Testnet(TestnetVersion::V3) => Magic::TESTNET,
+    Network::Testnet(TestnetVersion::V4) => Magic::TESTNET4,
+    Network::Signet => Magic::SIGNET,
+    Network::Regtest => Magic::REGTEST,
 }
 
 impl fmt::Display for Magic {
```

</details>

Reply:
> ```rust
> macro_rules! generate_network_magic_conversion {
>     ($(Network::$network:ident$((TestnetVersion::$testnet_version:ident))? => Magic::$magic:ident,)*) => {
>         impl From<Network> for Magic {
>             fn from(network: Network) -> Magic {
>                 match network {
>                     $(
>                         Network::$network$((TestnetVersion::$testnet_version))? => Magic::$magic,
>                     )*
>                 }
>             }
>         }
> 
>         impl TryFrom<Magic> for Network {
>             type Error = UnknownMagicError;
> 
>             fn try_from(magic: Magic) -> Result<Self, Self::Error> {
>                 match magic {
>                     $(
>                         Magic::$magic => Ok(Network::$network$((TestnetVersion::$testnet_version))?),
>                     )*
>                     _ => Err(UnknownMagicError(magic)),
>                 }
>             }
>         }
>     };
> }
> // Generate conversion functions for all known networks.
> // `Network -> Magic` and `Magic -> Network`
> generate_network_magic_conversion! {
>     Network::Bitcoin => Magic::BITCOIN,
>     Network::Testnet(TestnetVersion::V3) => Magic::TESTNET,
>     Network::Testnet(TestnetVersion::V4) => Magic::TESTNET4,
>     Network::Signet => Magic::SIGNET,
>     Network::Regtest => Magic::REGTEST,
> }
> ```

Thank you very much, your macro expression is very exquisite, I borrow your code to modify here